src/main/kotlin/com/refinedmods/refinedstorage/apiimpl/storage/disk/StorageDiskManager.kt

```diff
@@ -0,0 +1,89 @@
+package com.refinedmods.refinedstorage.apiimpl.storage.disk
+
+import com.refinedmods.refinedstorage.api.storage.disk.IStorageDisk
+import com.refinedmods.refinedstorage.api.storage.disk.IStorageDiskFactory
+import com.refinedmods.refinedstorage.api.storage.disk.IStorageDiskManager
+import com.refinedmods.refinedstorage.api.storage.disk.IStorageDiskProvider
+import com.refinedmods.refinedstorage.apiimpl.API
+import com.refinedmods.refinedstorage.extensions.Constants
+import net.minecraft.item.ItemStack
+import net.minecraft.nbt.CompoundTag
+import net.minecraft.nbt.ListTag
+import net.minecraft.server.world.ServerWorld
+import net.minecraft.util.Identifier
+import net.minecraft.world.PersistentState
+import org.jetbrains.annotations.Nullable
+import java.util.*
+
+
+class StorageDiskManager(name: String, private val world: ServerWorld) : PersistentState(name), IStorageDiskManager {
+    private val disks: MutableMap<UUID, IStorageDisk<*>> = mutableMapOf()
+
+    override operator fun get(id: UUID): IStorageDisk<*>? {
+        return disks[id]
+    }
+
+    @Nullable
+    override fun getByStack(disk: ItemStack): IStorageDisk<*>? {
+        val provider = disk.item
+        if (provider !is IStorageDiskProvider) {
+            return null
+        }
+        return if (!provider.isValid(disk)) {
+            null
+        } else get(provider.getId(disk))
+    }
+
+    override fun getAll(): Map<UUID, IStorageDisk<*>> = disks
+
+    override operator fun set(id: UUID, disk: IStorageDisk<*>) {
+        require(!disks.containsKey(id)) { "Disks already contains id '$id'" }
+        disks[id] = disk
+    }
+
+    override fun remove(id: UUID) {
+        disks.remove(id)
+    }
+
+    override fun markForSaving() {
+        markDirty()
+    }
+
+    override fun fromTag(tag: CompoundTag) {
+        if (tag.contains(NBT_DISKS)) {
+            val disksTag: ListTag = tag.getList(NBT_DISKS, Constants.NBT.COMPOUND_TAG)
```

Comment:
This will always return a new list. I think this is a bug from the original code. `getList` is implemented thusly:
```java
public ListTag getList(String key, int type) {
        try {
            if (this.getType(key) == 9) {
                ListTag listTag = (ListTag)this.tags.get(key);
                if (!listTag.isEmpty() && listTag.getElementType() != type) {
                    return new ListTag();
                }

                return listTag;
            }
        } catch (ClassCastException var4) {
            throw new CrashException(this.createCrashReport(key, ListTag.READER, var4));
        }

        return new ListTag();
    }
```

Reply:
Scratch that

Reply:
hehe, we all make mistakes sometimes =)

I actually had the same thoughts when reading the code, took me a while to figure out I was misreading it.